src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonIEnumerableWithAddMethodConverter.cs

```diff
@@ -0,0 +1,86 @@
+        where TCollection : IEnumerable
+    {
+        protected override void Add(object value, ref ReadStack state)
+        {
+            Debug.Assert(state.Current.AddMethodDelegate != null);
+            ((Action<TCollection, object>)state.Current.AddMethodDelegate)((TCollection)state.Current.ReturnValue!, value);
+        }
+
+        protected override void CreateCollection(ref ReadStack state, JsonSerializerOptions options)
+        {
+            JsonClassInfo classInfo = state.Current.JsonClassInfo;
+
+            if (classInfo.CreateObject == null)
+            {
+                ThrowHelper.ThrowNotSupportedException_DeserializeNoParameterlessConstructor(classInfo.Type);
+            }
+
+            state.Current.ReturnValue = classInfo.CreateObject()!;
+            state.Current.AddMethodDelegate = GetOrAddEnumerableAddMethodDelegate(classInfo.Type, options);
+        }
+
+        protected override bool OnWriteResume(Utf8JsonWriter writer, TCollection value, JsonSerializerOptions options, ref WriteStack state)
+        {
+            IEnumerator enumerator;
+            if (state.Current.CollectionEnumerator == null)
+            {
+                enumerator = value.GetEnumerator();
+                if (!enumerator.MoveNext())
+                {
+                    return true;
+                }
+            }
+            else
+            {
+                enumerator = state.Current.CollectionEnumerator;
+            }
+
+            JsonConverter<object> converter = GetElementConverter(ref state);
+            do
+            {
+                if (ShouldFlush(writer, ref state))
+                {
+                    state.Current.CollectionEnumerator = enumerator;
+                    return false;
+                }
+
+                if (!converter.TryWrite(writer, enumerator.Current!, options, ref state))
+                {
+                    state.Current.CollectionEnumerator = enumerator;
+                    return false;
+                }
+
+                state.Current.EndElement();
+            } while (enumerator.MoveNext());
+
+            return true;
+        }
+
+        internal override Type RuntimeType => TypeToConvert;
+
+        private readonly ConcurrentDictionary<Type, Action<TCollection, object>> _delegates = new ConcurrentDictionary<Type, Action<TCollection, object>>();
+
+        internal Action<TCollection, object> GetOrAddEnumerableAddMethodDelegate(Type type, JsonSerializerOptions options)
+        {
+            if (!_delegates.TryGetValue(type, out Action<TCollection, object>? result))
+            {
+                // We verified this exists when we created the converter in the enumerable converter factory.
+                result = options.MemberAccessorStrategy.CreateAddMethodDelegate<TCollection>();
+            }
+
+            return result;
+        }
+    }
```

Comment:
I realized this can be simplified since the converter should only be for a specific type. This diff shows a proposed change: https://github.com/layomia/dotnet_runtime/commit/39797f3ed94d7913577950598d72c2c1a0c08e4b?diff=split#diff-af03e9e4529096d4b0e7c9a55e6a06b0R72-R83.

Reply:
OK thanks I'll just apply it however you could add your own commit since we're not likely to collide.